### ISSUE TITLE
Android - Chat - User suggestion list is not opening on entering @ in 2nd line.

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#37660]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.46
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Launch app
2. Tap on a report
3. Enter @ in compose box
4. Select a contact from user suggestion list shown
5. Tap enter
6. Enter @ in compose box

## Expected Result:
User suggestion list must be opened on entering @ in 2nd line

## Actual Result:
User suggestion list is not opening on entering @ in 2nd line. It's slightly opening and closed again

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/8e6dd5d1-987c-4918-b2b7-dad8a5a2f4a7

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lanitochka17** [comment]
@trjExpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp
CC @quinthar

**Comment by trjExpensify** [comment]
@lanitochka17 can we get the device and Android OS please? 

I'm trying to reproduce this on a Pixel 6, Android 12 using Browserstack and can't seem to. Thanks!

**Comment by lanitochka17** [comment]
@trjExpensify Hello
Redmi note 10s Android 13

**Comment by trjExpensify** [comment]
Got it. Is it device specific? I don't have Redmi note 10s in Browserstack. I've tried in addition with Redmi 12 pro (Android 12) and Redmi 12 4G (Android 13), both of which I can't reproduce. 

Can we have it re-tested and also use a couple of different Android devices please, thanks!

**Comment by trjExpensify** [comment]
Awaiting the retest.

**Comment by trjExpensify** [comment]
Thanks!

**Comment by trjExpensify** [comment]
Ah, I missed the `External` label.

**Comment by DrLoopFall** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User suggestion list is not opening on entering @ in 2nd line.

### What is the root cause of that problem?
Typing the `@` character (or any character on empty line) triggers a scroll event.
Currently, we are hiding the suggestion list when the user scrolls the text input.
But typing text also generates scroll event, so we are debouncing hiding suggestion list on scroll event using `isScrollLikelyLayoutTriggered`,
https://github.com/Expensify/App/blob/022db8f5be1264e9fb45a7393c84ab2d309a97af/src/pages/home/report/ReportActionCompose/ReportActionCompose.tsx#L128-L141
using [useDebounce](https://github.com/Expensify/App/blob/82b76c1a36cf1a01b00c14510a5812a591e7f2fd/src/hooks/useDebounce.ts#L22)

The problem is in `useDebounce` function,
https://github.com/Expensify/App/blob/82b76c1a36cf1a01b00c14510a5812a591e7f2fd/src/hooks/useDebounce.ts#L24-L27

As we can see, we are passing `lodashDebounce` with option `maxWait` as `undefined` (if not provided), which is not same as not passing the `maxWait` at all, see https://github.com/lodash/lodash/issues/4224 
So currently, the `useDebounced` function acts as `throttle` function (https://lodash.com/docs#throttle)

So whenever the user types text, we are calling `raiseIsScrollLikelyLayoutTriggered`, which sets `isScrollLikelyLayoutTriggered` to `true`, which sets itself to `false` after some delay, using debounced, but it's not actually debounced but throttled, so it sometimes sets itself to `false` without the delay, which causes the suggestion list to hide. 

### What changes do you think we should make in order to solve the problem?
To solve this issue, we should remove `maxWait` which is actually not used anywhere.
We can also remove `leading`.

### What alternative solutions did you explore?

We can pass the `lodashDebounce` `options` argument without `maxWait` if `maxWait` is `undefined`.

**Comment by jayeshmangwani** [comment]
Tried with,
1. realme 11 device and Android version 13 (physical device)
2. Pixel 5 emulator with Android version 13 (emulator)

I can't seem to reproduce this issue on both the devices

**Comment by jayeshmangwani** [comment]
@DrLoopFall On which device(s) are you able to reproduce this issue ?

And also a question from the [proposal](https://github.com/Expensify/App/issues/37660#issuecomment-2004719022), why issue reproduces on some android devices and not on others ?

**Comment by DrLoopFall** [comment]
I can reproduce it in Android 13 (emulator)

After checking it once again, looks like this issue can be reproduced in 2 ways.
1. by typing fast (500ms delay between `newline` and `@`)
 
https://github.com/Expensify/App/assets/120749263/8af94900-5f4e-4926-83c9-f2412637da4d

2. by using a slow device (can be emulated using an Android emulator on a system with high CPU usage/some background task running), can also be reproduce in high traffic account with some devices.

https://github.com/Expensify/App/assets/120749263/f0b26f60-d361-4035-b3a7-da6583ac56c0

**Comment by DrLoopFall** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/37660#issuecomment-2004719022)

Added solution for **`issue #2`**

**Comment by jayeshmangwani** [comment]
@DrLoopFall If we conditionally render the `maxWait`  and `leading` or we can say if we remove these options altogether, then this issue will solved on the slow devices, right ?
then why we need to use the lodashDefer, and lodashDefer is simply the timeout under the hood if i am not wrong.

**Comment by DrLoopFall** [comment]
@jayeshmangwani
No, removing `maxWait` won't solve the issue for slow devices, it only solves the `issue #1`(fast typing).

Yes, `lodashDefer` is just `setTimeout` with a 1ms delay, it allows the other queued-up events to be triggered before `lodashDefer` function is called.
The `issue #2`(slow device) occurs because the scroll event is delayed when the JS thread is taking a long time to respond, so the scroll event is fired after the debounce function is called, so using `lodashDefer` allows the queues-up scroll event to be triggered before setting `isScrollLikelyLayoutTriggered`, which solves the `issue #2`

**Comment by jayeshmangwani** [comment]
@DrLoopFall Thanks for the thorough investigation,
I may agree on adding the conditional **maxWait** prop, but for adding the **lodashDefer** for solving the issue that is not consistent with the all the devices is not seems right to me, as I still think here is that we're just adding the **timeout** and that may solve the issue but thats not solving the root cause 

lets wait for today if any other may have proposal to solve this issue,otherwise I'll open the discussion in #expensify-open-source for adding the lodashDefer to setting  isScrollLikelyLayoutTriggered value

@DrLoopFall And a question, are you able to reproduce in any emulator ?

**Comment by DrLoopFall** [comment]
> @DrLoopFall And a question, are you able to reproduce in any emulator ?

Yes, I already mentioned it here, see https://github.com/Expensify/App/issues/37660#issuecomment-2007948188


> I may agree on adding the conditional **maxWait** prop, but for adding the **lodashDefer** for solving the issue that is not consistent with the all the devices is not seems right to me, as I still think here is that we're just adding the **timeout** and that may solve the issue but thats not solving the root cause
> 
> lets wait for today if any other may have proposal to solve this issue,otherwise I'll open the discussion in #expensify-open-source for adding the lodashDefer to setting isScrollLikelyLayoutTriggered value

Actually `isScrollLikelyLayoutTriggered` is made just to solve the same issue, specifically for Android (issue #26197 PR #27490), so I don't see any problem using `lodashDefer` to solve this issue, because it's specific for Android only.

**Comment by jayeshmangwani** [comment]
> Yes, I already mentioned it here, see https://github.com/Expensify/App/issues/37660#issuecomment-2007948188

Strange thing is I tested with Android 13 emulator too, but everything working fine, even though fast typing is working as expected

**Comment by DrLoopFall** [comment]
Maybe we should open a discussion in Slack, so others can also test it.

**Comment by jayeshmangwani** [comment]
Opened the discussion for testing the output in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q/p1711100869842639)

cc: @DrLoopFall @trjExpensify

**Comment by trjExpensify** [comment]
Doesn't look like anyone in the thread can reproduce it on the devices they've used. @shubham1206agra thinks it might be something to do with Live Markdown, but it's a bit TBD on why only certain devices are impacted.

**Comment by jayeshmangwani** [comment]
@trjExpensify Contributor has mentioned here that [issue](https://github.com/Expensify/App/issues/37660#issuecomment-2007948188) is reproducible on the slow devices, though I have tested on 2 device and both worked well both has Android version 13, so not sure whats the next step here, should we try to fix to fix this issue for the some devices ?

**Comment by trjExpensify** [comment]
Do we have an idea of the root cause? As in, is it an issue on all devices under certain conditions?

**Comment by jayeshmangwani** [comment]
@trjExpensify No, issue reproducible conditions are vague, as issue reproducible for some devices (not for the exact android version or particular mobile brands), and to be honest I haven't faced issue in any of the devices that I have tested recently

**Comment by DrLoopFall** [comment]
Hi @cubuspl42, @Gonals, @bernhardoj
Since you've worked on same issue #26197 (User suggestion @ list not appearing on the second line)

Can you please check if this issue is also reproducible?

---

**Here are the steps to reproduce the issue:**

1. by typing fast (500ms delay between `newline` and `@`)
 

<details>
<summary>Screenrecord</summary>

https://github.com/Expensify/App/assets/120749263/8af94900-5f4e-4926-83c9-f2412637da4d

</details>

 2. by using a slow device (can be emulated using an Android emulator on a system with high CPU usage/some background task running)

<details>
<summary>Screenrecord</summary>

https://github.com/Expensify/App/assets/120749263/f0b26f60-d361-4035-b3a7-da6583ac56c0

</details>

**Comment by bernhardoj** [comment]
I can reproduce it.

https://github.com/Expensify/App/assets/50919443/8d422587-fb24-4e95-a5a9-b628daad669c

**Comment by jayeshmangwani** [comment]
Thanks for the checking @bernhardoj 
As you can reproduce this issue, it maybe helpfull if you can share the emulator details, so that I create the emulator with same configuration, then test and reproduce it reliably, i have tried quite few emulatoors but suggestion modal working fine for me, are you able to reproduce this issue everytime relaibly ?

**Comment by bernhardoj** [comment]
I use Pixel 4a API 33, I never changed the settings, but here you go:

<img width="552" alt="image" src="https://github.com/Expensify/App/assets/50919443/f4897c13-110f-47bb-a591-11579a2d762d">
<img width="549" alt="image" src="https://github.com/Expensify/App/assets/50919443/e978044e-301e-4f50-907c-88f07cdd07d7">
<img width="562" alt="image" src="https://github.com/Expensify/App/assets/50919443/1f34defe-6563-4229-8cb2-c32e353b5bbe">

**Comment by jayeshmangwani** [comment]
Thanks for the help @bernhardoj , finally I was able to reproduce this issue using [this emulator](https://github.com/Expensify/App/issues/37660#issuecomment-2029125312).

**Comment by jayeshmangwani** [comment]
@DrLoopFall's [Proposal](https://github.com/Expensify/App/issues/37660#issuecomment-2004719022) of passing the conditional leading and maxWait props and using **lodashDefer to useCallback** will solve this issue for the slow devices. Compared before and after applying the lodashDefer in Pixel 4a, it solves the current issue

🎀 👀 🎀 C+ reviewed

**Comment by trjExpensify** [comment]
@madmax330 you good with [this proposal](https://github.com/Expensify/App/issues/37660#issuecomment-2004719022)?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by trjExpensify** [comment]
>Oh, interesting. Can you guys still reproduce?

Sorry, this was directed to @DrLoopFall & @jayeshmangwani! Let me know and we can close here if it's resolved.

**Comment by jayeshmangwani** [comment]
@trjExpensify From the latest Play Store build, I was able to reproduce this issue only once after testing it for 10-15 minutes.

I am not able to reproduce this issue in the latest main branch; before, I was able to reproduce it easily from my local build, but now the problem is not happening to me.

**Comment by DrLoopFall** [comment]
@trjExpensify,  I'm also unable to reproduce this issue in the main branch.
I think we can close this now.

**Comment by trjExpensify** [comment]
Okay, cool. Let's close it out!
